### ISSUE TITLE
Licensing

### ISSUE THREAD

**Issue by aboodman** [rocicorp/replicache#598]
In order to implement our billing model, we need to track the number of unique clients each customer uses Replicache with. 

This implies that Replicache will ping some central server on startup with an accountid/clientid pair so that we can count. Replicache generates the clientid internally already, and the accountid could be supplied as a constructor parameter.

However, we do **not** want it to be possible to accidentally (or maliciously) use someone else's account id  and charge their credit card. Therefore, accounts should be somehow tied to domains -- and you should only be able to construct Replicache with an accountid if the owner of that account intends Replicache to be used on that domain. (As a nice side effect this would allow account holders to know which domains are generating Replicache usage).

Another requirement is that we do **not** want to report domain names Replicache is used on to some central service. It's common for companies to have internal, semi-secret host names, and it would be bad practice to collect those.

**Comment by aboodman** [comment]
Strawman:

* There is an npm script in the Replicache package that generates a *license key* which encodes:
  - a unique account id
  - a single host (by default, perhaps localhost?) that Replicache can be instantiated on
* The license key is signed by a private key known only to Rocicorp
* The license key and signature are provided as arguments to Replicache at startup
* Each time Replicache is constructed it validates the provided license key signature using Rocicorp's public key (embedded into Replicache) and also that the current host matches the allowed host
* If the license and usage is valid, Replicache pings a central server with:
  - The hash of the license key
  - The account ID from the license key
  - The client ID Replicache is instantiated in
  - (maybe?) the ordinal of the host in the list of allowed hosts (always zero with this default license key that only allows one host)
 * The server records the ping associated with the account and client
 * Once a month some other process uses this data to charge customers

**Comment by aboodman** [comment]
Note: since this bug was filed the `clientID` in Replicache has become local to a single tab/session. So we need another permanent `profileID` to track unique profiles as opposed to clients.

Maybe it would be wise to report both `profileID` *and* `clientID` so that we could have flexibility on pricing model in the future.

**Comment by phritz** [comment]
> we do not want it to be possible to accidentally (or maliciously) use someone else's account id and charge their credit card ... Therefore, accounts should be somehow tied to domains

I do not think we should have this as a requirement, or at least we should not have this as a requirement right now. This feels like one of those requirements we ultimately did away with in DD like "a snapshot should be useful without first having to hit disk": sure, _ideally_ it's not possible to accidentally charge else's account, but practically speaking it doesn't seem worth the effort to implement such a mechanism at this point. We don't have this problem and adding the mechanism now constrains the way Replicache works in ways potentially annoying to customers (they go to use Replicache on some new thing and it doesn't work).

I do not think we need to do anything about misuse of keys at the moment, but if we must how about instead of _preventing_ the misuse of keys we instead provide tooling that makes it easy to _detect_ and _recover from_ misuse of keys? There are lots of obvious ways we can do this when the time comes and it seems easier/less annoying than prevention. 

I definitely think we should have some lightweight mechanism that makes accidental misuse of licenses less likely, but I do not think it should be a requirement to _prevent_ it, and I definitely do not think Replicache should stop working if you start using it on an unexpected domain.

**Comment by phritz** [comment]
BTW regarding a customer cloning our repo and using whatever key happens to be in there, that's the kind of accidental misuse that I think we should have a lightweight mechanism to catch, and I think that that lightweight mechanism can be a lot less work to implement and a lot less onerous for customers than what was suggested above.

> I definitely think we should have some lightweight mechanism that makes accidental misuse of licenses less likely, but I do not think it should be a requirement to prevent it, and I definitely do not think Replicache should stop working if you start using it on an unexpected domain.

**Comment by aboodman** [comment]
Hm maybe. Make a counter-proposal?

**Comment by aboodman** [comment]
I guess I agree it's not a *requirement* right now. But it seems like it will quickly become an issue especially with people accidentally copying the sample license key. I'm not sure how we could detect / understand that if we can't tell where the usage is coming from.

**Comment by aboodman** [comment]
> I think that that lightweight mechanism can be a lot less work to implement and a lot less onerous for customers than what was suggested above.

Interested to hear what you have in mind!

**Comment by phritz** [comment]
> I guess I agree it's not a requirement right now. But it seems like it will quickly become an issue especially with people accidentally copying the sample license key

I don't think we have to treat those two separate problems as one problem. I certainly agree that we should do something right now to prevent people from using the sample license key for something other than tire-kicking. However I don't think we need to prevent customers from using other customers keys right now. The first problem can be solved in a variety of simple ways that don't require elaborate mechanisms, for example if the URL contains "replidraw" then they are probably tire-kicking, if not we tell them to get a key.

I will make a proposal.

**Comment by aboodman** [comment]
> constrains the way Replicache works in ways potentially annoying to customers (they go to use Replicache on some new thing and it doesn't work because domain).

Unsurprisingly, I'm sympathetic to this bit.

> if the sample key is used we can check in the client if the URL contains "replidraw"

I think there are other variants of this "sample app problem" -- when people create and share tutorials online on their own blogs, what sample key will they use? Maybe the sample keys should only run for a short time or something, like in minutes, then you have to reload the app?

**Comment by aboodman** [comment]
Open to ideas like this as long as people don't accidentally end up using sample keys and people can still create and share tutorials. (as usually there are probably other requirements/desires in my head that we won't find until we start discussing alternatives).

**Comment by phritz** [comment]
Proposal: https://docs.google.com/document/d/1MxPhS55ie57TdjSPfrq8B5xQCug9hJW1GhpKH1tD9lA/edit?usp=sharing

Sorry/not sorry it's in a doc, we can make public or highlight essential elements in the issue if it is of interest outside Rocicorp.
